### ISSUE TITLE
Setting an achievement (A) as a step to earn another one (B) implies a limit on max earning for (A)

### ISSUE THREAD

**Issue by Sym-O** [opencredit/badgeos#458]
Hi ! 
I'm encountering a weird issue now. I found out that at some point an achievement stopped to be awarded whereas it has been attributed by nomination to a user regardless to the max earnings setting.
- after avoiding the #380 problem by setting max earnings to 100 instead of 0, the problem was still there, my achievement still wasn't earnable more than 4 times. It was working with an admin attribution, but not nomination.
- after some search, I found out that it seems to be related to the parent relation. This specific achievement was a step to earn another badge. We needed 4 of this one to unlock one step of another achievement. I played with it a bit to understand.
- moving this limit to 5 let us attribute one more achievement
- adding a new badge just with a need of Two earnings of the first achievmeent let us give one more achievement. It's just the "time" badgeos understand that a new step is succeed. After that no more nomination possible for badge (A).
- I tried to play with the object "Step" created for the link between (A) and (B). By default the max earnings settings of steps is 1. I tried to increase it to 5. It does NOT change things.
- I tried the same without to much conviction for the achievement (B). Nothing.
Could you help me please ?

**Comment by tw2113** [comment]
Could you provide all of the steps and details of each one? I'm trying to mentally visualize and am struggling.

**Comment by Sym-O** [comment]
OK ! 
Achievement "X" has 2 steps : 
- Earn 4 times Achievement "y"
- Earn 1 time Achievement "z"
Max earnings for "y" is set to 100 to avoid the Impossible Zero issue.
The issue was that it was impossible to nominate a user to receive more than 4 times the achievement "y" whereas an other achievement "u", completely unrelated, was givable as many time as the user was nominated. 
It is still possible to attribute "y" in admin mode, but not with nomination. 
A front nomination on "y" correctly create the nomination and when it is approved, an email is sent but no real attribution is made. User stay at 4 achievement "y".

If I set "y" to be a step to another unrelated achievement "V" (for exemple earn 3 times "y"), the next nomination of "y" on the same user will work one time, bringing the total of "y" to 5 and then the step of "V" will be completed and then again, no more nomination is possible.

The tests leads almost for sure to a problem in the nomination validation with the parent of the achievement. But I don't see exactly where is the problem. 
Is it more clear ?

**Comment by tw2113** [comment]
What about details for achievement Y and Z ?

**Comment by Sym-O** [comment]
"y" and "z" are both earn by nomination form. Is it the meaning of your question ?

**Comment by tw2113** [comment]
Yeah, any details that would be useful in potentially recreating on my own.

**Comment by Sym-O** [comment]
I think that's it. 
- Create 3 achievement (in my case the father X is from one achievement type (badge) and the 2 sons y and z are from another one (success))
- X is earned by steps : 4 y and 1 z
- y and z are earned by nomination
- Max earnings for X is 1, for y is 100, for z is 100
- do the earnings using the nomination form in the front office
That should do it.

**Comment by tw2113** [comment]
Hrm. At least with my local setup, I can't seem to reproduce well enough to be useful. I'm actually failing a lot earlier, because I can't get past more than one nomination for the achievement. Specifically in the `badgeos_create_nomination` function. I don't get past the `badgeos_check_if_user_has_nomination` check, which needs to return false. I keep getting true.

**Comment by Sym-O** [comment]
Arf ! I may have had the same issue. Double check the achievements and the field max earnings. You could even go to the database to be sure that this field is well set in the post_meta table for each achievement. 
I haven't done enough testing to be sure, but with this issue of being not able to exceed 1 attribution, whereas you set things otherwise, I feel I have encounter weird behavior...

**Comment by tw2113** [comment]
The settings should be valid, because I did a reset of my local install before setting up based on the info provided above.

Unrelated, but would you also be willing to help test what will hopefully be BadgeOS 1.4.5? All bugfixes. You can grab a copy at https://github.com/opencredit/badgeos/tree/1.4.5. Changelog: https://github.com/opencredit/badgeos/blob/alpha/readme.txt#L223

**Comment by Sym-O** [comment]
Hi ! 
Yes I'll try to integrate our data in the 1.4.5 to see how it behaves. 

Did you succeed to attribute more than one achievement ? That's strange you can't where as I can... I don't remember having implemented a fix for that... ^^

**Comment by tw2113** [comment]
1.4.5 isn't going to have anything knowingly related to this bug at hand. It just has a number of fixed bugs built up over time.

**Comment by tw2113** [comment]
I never got my local testing to award 2 of the same badge to someone. Only variable I can think of here would be that I can't award myself multiple times. I didn't test with a 2nd user in the install.

**Comment by Sym-O** [comment]
About 1.4.5, I did understand that it wasn't related to this topic. 

About this topic, that's strange, because I can award myself multiple time by nomination... Don't know how to help you move forward on this. 

Maybe, If you already have set the achievement you're trying to award yourself as a step for another achievement : Try to create a complety unrelated achievement awardable by nomination and set its max earnings to 100. Check if the 100 is well saved. and then try to award yourself that new achievement a few times.

**Comment by Sym-O** [comment]
Don't know if you move forward on this but I just did. 
In fonction badgeos_user_has_access_to_achievement, a check is made recursively on the parent of an achievement. If one of the upper achievement can't be given, the current achievement can't be. What took me time to understand fully the situation is that there is a hook to this function to handle the "step" case. If a step has already been earned the process is simply blocked no matter what is the value of its max earnings setting. (it's a bit strange here but can be explained that the steps are automated built achievement)
According to me there is an issue here but we have various approach to solve it. 
The one I choose for now is to disable the recursive check on the parent. If I specificaly set an achievement to be earnable an unlimited amount of time, it's not for him to be blocked by it parent. The actual behavior can go against the admin will. That's not good. 
But we could also discuss about the limitation to 1 step which is a problem. For example, if I have an achievement eanable by steps and I want it to be earnable 2 times or more. It won't be possible because I won't be able to cumulate more than one of each of his steps.

**Comment by tw2113** [comment]
Makes sense. Sorry for sort of dropping off the radar with this one. I haven't given up on it.
